### ISSUE TITLE
Add path to global.json SDK version lock

### ISSUE THREAD

**Issue by rainersigwald** [dotnet/sdk#8254]
Builders of large systems often want to be able to explicitly specify an SDK version _and_ avoid installing the CLI to a machine-global location.

Currently, `global.json` allows the version lock, but the specified version must be installed in either a machine-global location or a nonstandard location must be specified by the environment variable `DOTNET_MSBUILD_SDK_RESOLVER_SDKS_DIR`.

That environment means that you must launch Visual Studio from a specific environment to get the downloaded/private SDKs.

There could be an extension to the SDK resolver to respect a path specified in the `global.json`. Something like

```json
{
    "sdk": {
        "version": "1.0.0",
        "path": "tools/downloadedsdk"
  }
}
```

could be equivalent to setting `DOTNET_MSBUILD_SDK_RESOLVER_SDKS_DIR=%GlobalJsonPath%\tools\downloadedsdk` before invoking the resolver.

This would be visible by any invocation (`dotnet`, `msbuild`, VS, or MSBuild API) since it's file-based.

**Comment by abelbraaksma** [comment]
This would be a very welcome addition to the rather inflexible `global.json` config options. Currently, not being able to point to a local path, and the VS installer constantly uninstalling previous SDK versions, makes it pretty hard to do any decent SDK-dependent development (like the F# repo project files, see dotnet/fsharp#10193, which should be fixed on 3.1.302, but the VS installer removes this in favor of 3.1.402). 

Since it is trivial to create an msbuild script that downloads a specific SDK version and installs it locally, it would be _very, very nice_ if such behavior could be baked into `global.json` as well:

* specify a local path, or just use token `"local": true` and standardize the path to be `.dotnet`
* if absent, VS installs the SDK automatically upon fist opening the solution

All in all, that would create a much better experience then the now rather hard-to-understand "cannot load project or solution" message, which baffles newcomers and experienced users alike...

**Comment by RussKie** [comment]
This will have a significant positive impact on the developer experience for .NET repositories, such as https://github.com/dotnet/runtime/, https://github.com/dotnet/winforms, etc. Currently it is impossible to double click on a solution file to open the solutions in VS, because we build against nightly SDKs, which are located in custom locations and are forever chaning.

**Comment by KirillOsenkov** [comment]
As a workaround, I keep around a .bat file with this example content:

```
set PATH=C:\msbuild\.dotnet;%PATH%
set DOTNET_INSTALL_DIR=C:\msbuild\.dotnet
set DOTNET_MULTILEVEL_LOOKUP=0
```

If you go to the repo directory (C:\msbuild in my case) and run this .bat file, then run devenv.exe mysolution.sln, it will open with these environment variables set and the SDK resolution by MSBuild will use the .dotnet that's private to the repo.

**Comment by RussKie** [comment]
That's what other dotnet repos are doing AFAIK in some shape or form. And this is not the best developer experience. It also makes it hard to run ad-hoc tests...

**Comment by KirillOsenkov** [comment]
Oh, you don't have to convince me ;) I know it's terrible and I've spilled blood, sweat and tears to even arrive at this workaround by debugging deep into the SDK resolution process.

Lots of pain and friction caused by this over the years.

**Comment by RussKie** [comment]
To include other points form offline chats and summarise:

There a number of use cases that really benefit from this (or comparable) functionality:

  1. Developer experience building and contributing to .NET repos, such as https://github.com/dotnet/runtime/, https://github.com/dotnet/winforms, etc.  Currently it is impossible to double click on a solution file to open the solutions in VS, because we build against nightly SDKs, which are located in custom locations and are forever changing.  This make it unnecessary hard and complicated for our (especially new) contributors. Right now a lot of our tooling is cli-centric (which may be expected for non Windows platforms and CI/CD scenarios), but our Windows/Visual Studio devex is very cumbersome. Repos have custom scripts or instructions that must be run in order to open solutions in Visual Studio.
  2. Local testing - it is very difficult to run customer repos against the nightlies. These builds must either be installed globally, or one has to jump through hoops and run custom scripts to bootstrap a sample.

If it is difficult to provide the download functionality, we could probably do this in staggered approach, i.e. provide the ability to resolve an SDK from a custom location. 
* If the path is not specified - continue doing what we do today. 
* If the path is invalid - fail.

Maybe add another property to specify an error message that can describe how to bootstrap or where to download from. E.g.:
```
{
  "sdk": {
    "version": "2.1.100-preview-007366"
  },
  "msbuild-sdks": {
    "RoslynTools.RepoToolset": { 
      "version": "1.0.0-beta2-62705-02",
      "path": ".sdk",
      "instructions": "Run .\restore.cmd to download the latest SDK.",
    }
  }
}
```

**Comment by marcpopMSFT** [comment]
@vitek-karas We explored an option offline of just supporting this within the sdk resolver to find the SDK (essentially just DOTNET_MSBUILD_SDK_RESOLVER_CLI_DIR). Doing only that would unblock build but it would leave run (and potentially testing) still blocked as teams like WindowsDesktop have a test app that they would want to build and launch from VS against a version of the runtime listed in the global.json file.

Thoughts on finding the runtime from a path in global.json as well? I think the scenario for dotnet repos would have the runtime and SDK all in the same directory so potentially we only need one path still for both.

**Comment by RussKie** [comment]
To add to @marcpopMSFT's post.

Starting VS without settings any paths or anything:
![image](https://user-images.githubusercontent.com/4403806/123887529-04464280-d995-11eb-81f4-926dace7c096.png)

Starting VS with `env:DOTNET_MSBUILD_SDK_RESOLVER_CLI_DIR` settings:
```powershell
PS C:\Development\winforms> $env:DOTNET_MSBUILD_SDK_RESOLVER_CLI_DIR='C:\Development\winforms\.dotnet'
PS C:\Development\winforms> start .\Winforms.sln
```
![image](https://user-images.githubusercontent.com/4403806/123887577-1f18b700-d995-11eb-93fa-3803f36428ce.png)
On a second F6 it all build successfully... ¯\\_(ツ)\_/¯

But I was unable to launch a project from the VS:
![image](https://user-images.githubusercontent.com/4403806/123887704-54250980-d995-11eb-8fbb-c0c5ed24c990.png)

**Comment by vitek-karas** [comment]
From runtime's perspective there's a risk of performance regressions. This would require the host to go looking for `global.json` (or some other similar file) performing file existence checks. The deeper the folder hierarchy, the more expensive the check. This would have to be done by EVERY application, EVERY time. Given that this feature is expected to be used almost exclusively by developers that doesn't feel like the right tradeoff.

Doing this for SDK only is in theory possible, but could open issues with compatibility between SDK and the necessary runtime/frameworks for it (since we could not rely on the `dotnet.exe` to be the right one anymore).

I absolutely understand the scenario and why it's important, but so far I was not able to figure out a solution which would not have a negative effect when the feature is not is use. (pay-for-play)

/cc @agocke

**Comment by marcpopMSFT** [comment]
@RussKie Thoughts on only supporting this for the SDK as originally suggested given Vitek's feedback above?  It wouldn't allow for launching test apps from the IDE but would allow for building. It's unclear if tests would work for not.

**Comment by danmoseley** [comment]
@agocke thoughts? Seems there's agreement that it would be high value to solve this as it will make us and the community more productive in our repos.

**Comment by DustinCampbell** [comment]
I (think) I like this. :smile: However, pinning the SDK in global.json is already a significant problem within Visual Studio, due to mismatches between the MSBuild engine in Visual Studio and the targets/tasks loaded from the .NET SDK. I have some worry that this might exasperate that problem more.

@KathleenDollard, do you have any thoughts?

**Comment by richlander** [comment]
Is it possible to start with an ENV solution only? That would mitigate Vitek's concerns. I'd like to reduce the number of places we have to search on disk not invent more.

**Comment by agocke** [comment]
Having repo-local SDKs make sense and I think is tractable. The team talked about repo-local runtimes and that one seems much more difficult. Not only would it probably be a big perf problem, as more intermediate tools use .NET Core (I hope), it's less and less likely that you want *everything* to run on local runtime. For instance, would you want new VS components that run on .NET Core to suddenly pick your repo local runtime?

I think SDK is a good place to start, if we can make that work.

**Comment by richlander** [comment]
If we had a repo root story, then putting a global.json at repo root would be workable and I suspect mitigate Vitek's concerns (depending on how repo-root worked; if it was an ENV, it would work awesome).

**Comment by RussKie** [comment]
> @RussKie Thoughts on only supporting this for the SDK as originally suggested given Vitek's feedback above?  It wouldn't allow for launching test apps from the IDE but would allow for building. It's unclear if tests would work for not.

This wouldn't work for us. I can build from command line and don't need VS for that.
I'd like a seamless devex for VS: double click sln, F5, debug.

**Comment by RussKie** [comment]
RE: perf - we can make a call, and only look for .dotnet folder in the same folder as global.json.
We have many things that are convention based, we can probably add yet another thing 😊

**Comment by agocke** [comment]
@RussKie You're confusing the SDK with the runtime. `global.json` is an SDK configuration option. Right now there's no way to configure the muxer `dotnet.exe` to look somewhere else for the runtime aside from environment variables. Looking for `global.json` would run into the same problem -- anything you add to the muxer is overhead on _every execution of a dotnet app_. We've already removed a bunch of file system accesses from the host startup because they had unacceptable startup penalty. I don't consider adding any new filesystem checks to be an acceptable approach.

**Comment by Nirmal4G** [comment]
Then, why not do a context separation with `global.json` and other env checks?

If the file is found along with other env checks, then, the muxer should be in Dev/SDK mode; else it'll be in Runtime/Execution mode.

Do what you are doing today with the exec mode or simplify further if possible since every bit of perf is important.

But in Dev/SDK mode, do things that improves the developer productivity such as loading dev-specific options (_look for new SDK/Runtime paths, etc…_) from both the env variables and the `global.json` file.

Almost every project, if it needs customizability, should have a `global.json` in the root. So, we could reduce the checks to only look for `global.json` to determine the context.

**Comment by Nirmal4G** [comment]
Then, why not do a context separation with `global.json` and other env checks?

If the file is found along with other env checks, then, the muxer should be in Dev/SDK mode; else it'll be in Runtime/Execution mode.

Do what you are doing today with the exec mode or simplify further if possible since every bit of perf is important.

But in Dev/SDK mode, do things that improves the developer productivity such as loading dev-specific options (_look for new SDK/Runtime paths, etc…_) from both the env variables and the `global.json` file.

Almost every project, if it needs customizability, should have a `global.json` in the root. So, we could reduce the checks to only look for `global.json` to determine the context.

To add on further, you could have either `--mode <dev|run|…>` or `--dev` and `--no-dev` to force the muxer into a specific context.

**Comment by Nirmal4G** [comment]
Then, why not do a context separation with `global.json` and other env checks?

If the file is found along with other env checks, then, the muxer should be in Dev/SDK mode; else it'll be in Runtime/Execution mode.

Do what you are doing today with the exec mode or simplify further if possible since every bit of perf is important.

But in Dev/SDK mode, do things that improves the developer productivity such as loading dev-specific options (_look for new SDK/Runtime paths, etc…_) from both the env variables and the `global.json` file.

Almost every project, if it needs customizability, should have a `global.json` in the root. So, we could reduce the checks to only look for `global.json` to determine the context.

To add on further, you could have either `--mode <dev|run|…>` or `--dev` and `--no-dev` to force the muxer into a specific context.

**Comment by Nirmal4G** [comment]
Then, why not do a context separation with `global.json` and other env checks?

If the file is found along with other env checks, then, the muxer should be in Dev/SDK mode; else it'll be in Runtime/Execution mode.

Do what you are doing today with the exec mode or simplify further if possible since every bit of perf is important.

But in Dev/SDK mode, do things that improves the developer productivity such as loading dev-specific options (_look for new SDK/Runtime paths, etc…_) from both the env variables and the `global.json` file.

Almost every project, if it needs customizability, will have a `global.json` in the root. So, we could reduce the checks to only look for `global.json` to determine the context.

To add on further, you could have either `--mode <dev|run|…>` or `--dev` and `--no-dev` to force the muxer into a specific context.

**Comment by RussKie** [comment]
> For instance, would you want new VS components that run on .NET Core to suddenly pick your repo local runtime?

This is a good question. I think the answer is "it depends", with likely "yes".

**Comment by Nirmal4G** [comment]
> currently the muxer doesn't look for this file when running an app.

Sorry If I've got this all wrong. Doesn't it use different .NET SDK when we have a `global.json` with a particular `sdk` version mentioned? If it's not the muxer doing this, then what else?

> Running all my dev tools on top of such runtime

Not the case here. As a runtime or a dev tools contributor, I'd want only certain invokes of `dotnet` to be run in a developer context. Like testing the repo with a just built components of runtime (_e.g.: the SDKs, MSBuild or NuGet targets_) or in the case of core, winforms and wpf runtime, the entire runtime.

**Comment by vitek-karas** [comment]
The muxer (`dotnet.exe`) does two main things:
* Runs an application (`dotnet app.dll`, or `dotnet exec app.dll` for example)
* Invokes a CLI/SDK command (`dotnet build` or `dotnet run` for example)

When it detects (by looking at the command line basically) that it's to run CLI, it switches to the CLI mode, in which case it will go look for `global.json`. But if it's running in the "app" mode, it will not do that.
The discussion is about modifying the "app" mode to go search for `global.json`.

**Comment by Nirmal4G** [comment]
Do we require modifying the exec mode to look for `global.json`? Isn't that what the `runtimeconfig.json` is for?

What I was asking and hoping to achieve here is to extend `global.json` to select more than just SDK version and have a list of MSBuild SDK packages, to customize lookup/restore options for the MSBuild SDKs, and of-course the .NET SDK itself.

**Comment by RussKie** [comment]
> The discussion is about modifying the "app" mode to go search for `global.json`.

This is not _exactly_ what I had in mind.
I look at it from a layman point of view, and I was asking what can be done to make the developer experience contributing to .NET repos to the same level as building and debugging other (normal) .NET solutions, i.e. "_double click on a sln file, have VS resolve the required/expected version of SDK/runtime, F5_". By "VS" (in layman terms) I mean everything involved to build and debug an .NET app, the editor, the compiler, analyzers, MSBuild, etc.

**Comment by vitek-karas** [comment]
I understand the high-level scenario, but in my mind it basically boils down to some mechanism through which running a .NET app/command from the repo directory tree can be made to use repo-local SDK/runtime. Ultimately, if I build a test app from within the repo, and run it - it should get the repo-local runtime. Running tests, or using VS to run them from me, is just an extension of that I think.

If we were to scope this to only VS (for example), then there are more options, as we can (in theory) make VS understand repo-local things and run the tests/apps with that knowledge. But I think everything should work directly from command line as well. And the "F5" from command line is basically `/myrepo/artifacts/tests/mytest.exe` - and we want to make this use repo-local runtime.

(Note that `app.exe` is called `apphost` and it's almost the same code as `dotnet.exe` (called the "muxer") - so the discussion above about the "muxer" applies to the "apphost" as well).

**Comment by vitek-karas** [comment]
> Do we require modifying the exec mode to look for global.json? Isn't that what the runtimeconfig.json is for?

We could encode something into the `runtimeconfig.json`, but that would mean that all of the `runtimeconfig.json` files in the repo, would need to have this "special" thing in them. Which is problematic for official builds - those would have to turn it off (shipping apps would not want that), but then again testing the official build could not be done from within the repo... and so on.

Potentially we could encode something like this into the `runtimeconfig.dev.json` - but there's still the problem of official build (which would need to know to exclude that file).

**Comment by marcpopMSFT** [comment]
I don't think we can scope a solution to only VS and per an earlier comment, we cannot fix this with environment variables either (as we already have that solution in place and this feedback comes even with env variable options being in use). We stopped generating the .dev.json file by default. How would we make it generate it for local builds but not CI builds?

**Comment by vitek-karas** [comment]
I'm sorry I didn't make my previous comment clear. I don't think it's a good solution to encode something into the `.runtimeconfig.json` (or the `dev` version of that file). Both solutions have the problem that they require us to modify produced binaries, which is very problematic. It's also very possible such solution would not just work for testing either (since tests generate their own `.runtimeconfig.json`) - would need even more "weird" changes.

Just thinking out loud (it could be a terrible idea): New environment variable which would tell the hosts to look for `global.json` (or similar file) even for app invocation. We would require developers in affected repos to set this variable (probably globally on their machine) once. It would imply the perf impact of the host on pretty much all apps on the machine, but it would solve the problem of only affecting users which need this functionality. (There are potential improvements on this idea as well obviously).

**Comment by RussKie** [comment]
> We would require developers in affected repos to set this variable (probably globally on their machine) once.

I don't think this will work either, e.g. I work on multiple dotnet/* repos, with at least two - actively. Resetting an env variable every time I need to launch a repo in VS would be a nightmare.

Given that the issue has been opened since 2017, and we've determined that we can't make any practical changes to accommodate the seamless VS developer experience for dotnet/* repos, should we close this?

**Comment by richlander** [comment]
I am coming back to this issue after writing up a breaking change on multi-level-lookup. See: https://github.com/dotnet/core/issues/7131. I'm thinking that it will help this scenario significantly.

For the CLI experience, it's easy to use "C:\PrivateSDK\dotnet build" and end up using a global SDK when you want to use a private one. Super frustrating. Resolving that is the point of the break.

Going forward, I'd offer the following as best practice guidance for the private SDK scenario:

```bash
DOTNET_ROOT=C:\PrivateSDK
```

You could then type the following:

```bash
PS C:\Users\rich\fakerepo> C:\PrivateSDK\dotnet build
PS C:\Users\rich\fakerepo> .\bin\Debug\net7.0\fakeapp
App launched on .NET 7 from C:\PrivateSDK
```

If you wanted to avoid typing "C:\PrivateSDK" all the time, you could add the location to the path, with the following:

```bash
set PATH=C:\PrivateSDK;%PATH%
````

This scenario works super well if you can set these ENVs via a script or manually. It doesn't work at all if you cannot.

The original idea (I think) was to effectively expose the analog of `DOTNET_ROOT` in global.json. I think that's an interesting direction IFF we have a new scenario for Visual Studio that it needs to respect that (which includes testing). Alternatively, we could create a convention where Visual Studio looks for a `.dotnet` directory at repo root to achieve the same end. I actually prefer that. Again, I think that's a Visual Studio scenario, not a .NET one. Ideally, there would be a Visual Studio way to say "look in this repo-local directory for the dev platform" and it would work equally well for node.js (for example).

Am I on the right page (or even the right book)?

**Comment by vitek-karas** [comment]
The original issue is mainly around VS (at least that's my understanding). Specifically: clone a repo, maybe run `build` from root for the first time, double-click the `.sln` file in the root -> everything works.

That said, it would be really nice if we came up with a solution which works from CLI as well. For SDK, using `global.json` makes sense. There might be some technical challenges how to implement it, but I don't see any big architectural problems. But for runtime, so far we haven't come up with anything good. The obvious requirement is, that whatever we do here is 100% opt-in.

**Comment by richlander** [comment]
My take was that this was about getting VS to use a different toolset than the one VS ships (.NET SDK or MSBuild/Roslyn/NuGet that comes with VS). I think that's what you are saying, too. I think this scenario needs to come from the VS team, not the SDK team. It's primarily challenges/constraints in VS that precent this scenario from working. If they want to prioritize the scenario, we can certainly do our part.

**Comment by rainersigwald** [comment]
> It's primarily challenges/constraints in VS that precent this scenario from working.

I don't think that's true? If the .NET SDK resolver could find one based on a path in the `global.json` everything would work more or less fine, right?

**Comment by richlander** [comment]
I was referring to the core toolset (MSBuild, Roslyn, Nuget, Common T&Ts). Related to this conversation: https://github.com/dotnet/sdk/issues/8254#issuecomment-301249426.

I've heard time and again that global.json (independent of this topic) is not transparent to VS, that switching to an older or newer (or just different, given the strangeness of SDK bands) can break VS. We need to resolve that topic, right?

Zooming out, can we enable such a model transparently from VS, such that it is none-the-wiser?

Perhaps we're talking past one another. Happy to participate in a call on this to help us focus on the core issues.

**Comment by marcpopMSFT** [comment]
The customer feedback I've heard internally is specifically about wanting to clone a repo and double click the solution file in file explorer. That doesn't work today as you have to set environment variables first which we've been told is a barrier to entry for some customers. Hence the ask to put it in global.json so a repo owner can do it and the end developer doesn't have to know though that only works if they actually have the needed sdk installed...

**Comment by jaredpar** [comment]
If I have an SDK in hand why do I need to go through the trouble of setting an environment variable to use it? That means that it's a command line tool I can't reliably run unless I muck with some environment variables. 

I still strongly think that I should be able to use `dotnet build` to use the SDK that came with the exe to build my application. Maybe I need to provide a command line switch like `--this-one` but the solution should be in the CLI itself. It shouldn't require any weird pathing. 

Every other language framework in existing **except** .NET provides this solution.

**Comment by RussKie** [comment]
I'm currently OOF and writing from a phone.

My main motivation for this change is outlined in https://github.com/dotnet/sdk/issues/8254#issuecomment-860995649.

**Comment by davidwengier** [comment]
Just adding my vote to this, particularly https://github.com/dotnet/sdk/issues/8254#issuecomment-1023551867. If I'm understanding correctly, this would really help in the https://github.com/dotnet/razor-tooling/ repo, which currently requires setting a bunch of envars set. Having to ensure these are set before running VS is a pain, particularly if you're trying to use multiple versions of VS to verify things (switching between developer command prompts etc.), or if you're having to use multiple repos with similar set ups (as @RussKie mentioned too).

**Comment by richlander** [comment]
@vitek-karas and I are working on a spec that possibly includes what you want. Will share (very) soon. In any case, it will be a good starting point.

**Comment by richlander** [comment]
We backed a way a little from this. We decided NOT to put any new location content into `global.json`. It is too easy to trick developers into running malicious code. Instead, we're starting with a new CLI argument. VS* products can add support for this flag if they like. It won't happen automagically due to a .NET SDK feature.

**Comment by sandyarmstrong** [comment]
@richlander can you provide some more detail on the malicious scenario here? I don't understand how this is riskier than restoring nuget packages.

Adding this info to `global.json` seems like the best way forward for painless IDE support of custom SDK locations on a per-solution basis.

**Comment by richlander** [comment]
Today, you can clone a repo, build it, and assume that your globally installed SDK was used, provided by a vendor you trust. With this model, the `global.json` could point to a directory in the repo like `~/malicious-dotnet/` that could perform arbitrary operations on your machine as a consequence of running `dotnet build`. Certainly, NuGet packages can also do bad things in the build. We're aiming to reduce not increase security concerns.

I'm not suggesting that cloning and building an arbitrary repo is a safe thing to do today.

**Comment by akoeplinger** [comment]
Not sure I follow your argument, how is that different from having a .csproj like this:
```xml
<Project Sdk="Microsoft.NET.Sdk">
   <Target Name="Build">
      <Exec Command="rm -rf /home/richlander" />
   </Target>
</Project>
```

The assumption today is that if you run `dotnet build` then you're trusting the code.

**Comment by vitek-karas** [comment]
If I remember correctly the difference was that with the global.json path, it would be enough to run `dotnet --info` in the "wrong" directory.

**Comment by KirillOsenkov** [comment]
After wasting another couple of hours on this today, I am more convinced than ever that we should allow specifying a path to the SDK in the global.json (as opposed to the environment variables, which are terrible UX for discoverability and automation). When you open the .sln in Visual Studio, for example, it doesn't have the environment variables, and so the design-time build fails completely (you get errors like "System.Object not found").

The tooling could even automate downloading the SDK into that folder instead of having arcane Arcade tools do it.

I have been shouting from the rooftops for years that for every repo in the .NET world anyone should be able to `git clone && msbuild` and it should just work, without custom build.cmd, build.sh etc. The environment variables get in the way of that. I absolutely do not buy the security argument because it doesn't make the status quo any worse (any risks from building with a local SDK are already present when restoring and/or building with vanilla SDK as Alexander explains). Both the project source or any of the NuGet packages that are restored can be malicious.

Not to mention that the current status quo is _you run build.cmd_. And of course, it still builds with that custom SDK, which still entails the same alleged security risks.

By ignoring this problem we exclude a whole class of tools and automation from working out of the box. If any repo is buildable out of the box, you could run security analysis, semantic indexing for SourceBrowser or other code search tools, etc. etc. These tools can't read your build.cmd as it's totally opaque imperative code. Had we had declarative configuration in global.json (download  and use this exact SDK), it would open doors to automating analysis etc. Opening the .sln in VS would always "just work", whereas now it doesn't and causes errors in the design-time build that only a few of us know how to deal with.

I think the current situation is a major source of friction in the .NET ecosystem.

**Comment by xoofx** [comment]
Hey, just found this thread.

We are actually looking for a solution at Unity to have the SDK configurable by the global.json. Our scenario is that we will ship/lock-in a specific version of the .NET SDK/Runtime as part of the Unity Editor, and we need a way to resolve our shipped .NET SDK for the sln/csproj files that we will compile with MSBuild or that we will open from Visual Studio.

Having a way to override the SDK root path in global.json seems to be a good fit for our use case.

**Comment by tmat** [comment]
Re Arcade -- the purpose of Arcade is to unify build across dotnet repos by implementing as few additions on top of the shipping .NET SDK as possible to achieve that. That goal has been achieved. The Arcade SDK has been pretty stable for a while now. The next step should be to review the additions/workarounds that were introduced by Arcade SDK. Those that are generally useful, i.e. not specific to Microsoft build process should be productized in the shipping .NET SDK/msbuild and removed them from Arcade SDK. Downloading SDK would be one of these features.

Note that for testing purposes a repo might want to specify multiple versions of .NET runtime it wants to run tests against. See https://github.com/dotnet/aspnetcore/blob/0f517603dc0cf996f58ccd67fd12b513794de5a4/global.json#L5-L14

**Comment by tmat** [comment]
> every repo in the .NET world anyone should be able to git clone && msbuild and it should just work, without custom build.cmd, build.sh et

Do you mean that Desktop msbuild or `dotnet build` would check the version of the SDK specified in global.json and download one if it is not the one available on the machine and then run `dotnet build` from that matching SDK?

**Comment by RussKie** [comment]
@tmat - yes. Or as an easier alternative to emit an error message directing developers how to get the required SDK.

**Comment by hknielsen** [comment]
To add to @xoofx's comment, for the work we are doing with MsBuild in Unity this is one of our big workflow issues.
The only way i've found to handle SDK resolving to something specific is to add local nuget repo and pin the SDK version in the global.json.

I would really like Sdk Resolving to be much more flexible. Creating a SdkResolver instead of a path in global.json would maybe be better, but theres no way to have it automatically picked up across vs/dotnet/msbuild in a nice way. Or at least I havent found a way!

I would be super happy to handle the Sdk resolving myself, path or SdkResolver

**Comment by xoofx** [comment]
> Do you mean that VS msbuild or dotnet build would check the version of the SDK specified in global.json and download one if it is not the one available on the machine and then run dotnet build from that matching SDK?

If possible, I would actually prefer to split the requirement/concern of the automatic downloading of a SDK version in a separate issue and keep this issue for allowing to specify a specific location for a SDK. It seems that allowing to override the path would require a small change to the existing MSBuild Sdk Resolver (I don't know if it would require lots of change to VS), while downloading a version that is not installed and put it on the drive look more cumbersome to align (and where, and how to cleanup the cache and...etc.).

**Comment by simonferquel** [comment]
As VS already support downloading custom SDKs (and using them) and use the global.json file to know which version to download, I feel the change is not more complex than it is for dotnet build (which seem quite trivial from an external point of view, but I could be wrong). Maybe it uses the same code as dotnet CLI for implementing that ?

If the issue is accepted, I think someone in Unity Scripting team can implement it for dotnet CLI, and see if there is any additional work to do for VS.

**Comment by hknielsen** [comment]
@rainersigwald I can do the work, as we need to have a solution for moving to MSBuild.

Two possible solutions;
1. Path to an additional location for picking up SDKResolver`s, that will be loaded by the SDKResolverLoader.
2. Path that we can Resolve additional MSBuild SDK's from

Im ok with both, but I could see more flexibility with option 1.
Ie. if someone wanted and SDKResolver that downloaded the msbuild SDK from another location, they could create their SDK resolver that automatically would do that, and it would work in VS/MSBuild/Dotnet cli

**Comment by sharwell** [comment]
> We decided NOT to put any new location content into global.json. It is too easy to trick developers into running malicious code.
> ...
> Today, you can clone a repo, build it, and assume that your globally installed SDK was used, provided by a vendor you trust. 

@richlander This is not even remotely correct. The default use of NuGet allows for arbitrary code to be executed by dotnet restore, including the ability to bypass anything we might put on nuget.org for detection mechanisms by redirecting to a privately hosted feed. Please reconsider the basis for the argument, as teams are being unnecessarily penalized by this decision.

**Comment by KirillOsenkov** [comment]
I wish we had this support. Environment variables are causing us so much pain during local dev with a pre-release SDK. The need to set several environment variables and ensure your IDE also inherits them is a constant source of friction.

**Comment by jaredpar** [comment]
I've written up a design proposal that seeks to address the problems shared on this thread as well as others I've seen. Appreciate any feedback you all have on this. 

https://github.com/dotnet/designs/pull/303

**Comment by KirillOsenkov** [comment]
long sad story

**Comment by elinor-fung** [comment]
The [Provide SDK hint paths in global.json](https://github.com/dotnet/designs/blob/main/accepted/2025/local-sdk-global-json.md) design is implemented in .NET 10 Preview 3.

There are two optional properties:
```json
{
    "sdk": {
        "paths": [ ".dotnet", "$host$" ],
        "errorMessage": "The .NET SDK could not be found, please run ./install.sh."
    }
}
```
- `paths`: locations to use when searching for a compatible .NET SDK. Paths can be absolute or relative to the location of the *global.json* file. The special value `$host$` represents the location corresponding to the running `dotnet` executable. 
- `errorMessage`: custom error message to use when a compatible SDK cannot be found

Note that since this involved updates to both the runtime host and SDK, both the original installation (where `dotnet` is invoked) and the resolved SDK (that may now be found through the `paths` property) need to be .NET 10 Preview 3 or later.

**Comment by RussKie** [comment]
Fantastic! Thank you for getting it done 🎉

**Comment by akoeplinger** [comment]
> Do I understand it correctly that dotnet is not responsible for downloading the actual global.json SDK?

Correct, today that's handled in our repos by arcade scripts but there is discussion on creating a bootstrapper integrated into the sdk: https://github.com/dotnet/sdk/issues/50623

> Also, is .NET 10 Preview 4 installed globally is enough for my VS to see the bootstrap sdk from that path?

The `hostfxr` of the original host needs to be 10.0p3 or later and looks like VS 17.14 still ships with an older one right now.

**Comment by KirillOsenkov** [comment]
In terms of bootstrapping it's been a gap in the ecosystem, and is something that's for example preventing me from moving MSBuild Structured Log Viewer from netfx to modern .NET. I don't want to ship the runtime with the app, and I also don't want to expect the runtime to be installed.

There's an existing project that fills this gap:
https://github.com/Tyrrrz/DotnetRuntimeBootstrapper

I have it on my todo list to try it out, but I think it'll be great to have it built-in.

**Comment by RussKie** [comment]
FYI, packages from this individual are protestware, which may or may not be acceptable in context of each individual project.

**Comment by EgorBo** [comment]
Just for the reference, here is what Java folks often use for similar purposes: `jenv` https://github.com/jenv/jenv (I guess python's `venv` and ruby's `rbenv` are also related).

While dotnet/* repos seem to have some `build.cmd/sh` that does the heavylifting (although, kind of weird that instead of `dotnet build` one has to execute shell scripts) there are projects with `global.json` without such scripts, you're just expected to get the SDK yourself)

**Comment by agocke** [comment]
You can use my dnvm project as a bootstrapper: https://dnvm.net
